monitor: Display human-readable identities

pchaigno opened the pull request:
This pull request changes `cilium monitor`'s output to display human-readable identities instead of numbers for reserved (fixed) identities.

The output before this pull request resembles:

    <- host flow 0x0 identity 2->0 state new ifindex enp0s8 orig-ip 0.0.0.0: 192.168.33.12:31313 -> 10.11.0.113:40364 tcp SYN, ACK

After:

    <- host flow 0x0 identity world->unknown state new ifindex enp0s8 orig-ip 0.0.0.0: 192.168.33.12:31313 -> 10.11.0.113:40364 tcp SYN, ACK

Please see individual commit messages for details.

pchaigno committed:
monitor: Display human-readable identities

This commit changes cilium monitor's output to display human-readable
identities instead of numbers for reserved (fixed) identities.

The output before this commit ressembles:

  <- host flow 0x0 identity 2->0 state new ifindex enp0s8 orig-ip 0.0.0.0: 192.168.33.12:31313 -> 10.11.0.113:40364 tcp SYN, ACK

After:

  <- host flow 0x0 identity world->unknown state new ifindex enp0s8 orig-ip 0.0.0.0: 192.168.33.12:31313 -> 10.11.0.113:40364 tcp SYN, ACK

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
identity: Add unknown to the reserved identities

The main objective is to display a proper human-readable identity in
cilium monitor for the 0 (unknown) identity.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno commented on pkg/monitor/api/types.go:
```diff
@@ -22,6 +22,7 @@ import (
 	"strings"
 	"time"
 
+	"github.com/cilium/cilium/pkg/identity"
```
For reviewers: I'm unsure whether it's fine to import this package here.

joestringer replied:
Using https://github.com/KyleBanks/depth :

```
$ depth ./pkg/identity
github.com/cilium/cilium/pkg/identity
  ├ errors
  ├ fmt
  ├ net
  ├ strconv
  ├ sync
  ├ github.com/cilium/cilium/pkg/k8s/apis/cilium.io
  ├ github.com/cilium/cilium/pkg/labels
    ├ bytes
    ├ crypto/sha512
    ├ encoding/json
    ├ fmt
    ├ net
    ├ sort
    ├ strings
    ├ github.com/cilium/cilium/pkg/logging/logfields
      └ fmt
    └ github.com/cilium/cilium/vendor/github.com/sirupsen/logrus
      ├ bufio
      ├ bytes
      ├ context
      ├ encoding/json
      ├ fmt
      ├ io
      ├ log
      ├ os
      ├ reflect
      ├ runtime
      ├ sort
      ├ strconv
      ├ strings
      ├ sync
      ├ sync/atomic
      ├ time
      ├ unicode/utf8
      └ github.com/cilium/cilium/vendor/golang.org/x/sys/unix
        ├ bytes
        ├ encoding/binary
        ├ math/bits
        ├ runtime
        ├ sort
        ├ strings
        ├ sync
        ├ syscall
        ├ time
        ├ unsafe
        └ github.com/cilium/cilium/vendor/golang.org/x/sys/internal/unsafeheader
          └ unsafe
  └ github.com/cilium/cilium/pkg/lock
    ├ context
    ├ sync
    ├ sync/atomic
    ├ github.com/cilium/cilium/vendor/github.com/sasha-s/go-deadlock
      ├ bufio
      ├ bytes
      ├ fmt
      ├ io
      ├ io/ioutil
      ├ os
      ├ os/user
      ├ path/filepath
      ├ runtime
      ├ strings
      ├ sync
      ├ time
      └ github.com/cilium/cilium/vendor/github.com/petermattis/goid
        ├ bytes
        ├ runtime
        └ strconv
    └ github.com/cilium/cilium/vendor/golang.org/x/sync/semaphore
      ├ container/list
      ├ context
      └ sync
38 dependencies (28 internal, 10 external, 0 testing).
```

Seems like a lot but I'm guessing the callers *probably* already pull in many of these?

pchaigno commented on pkg/identity/numericidentity.go:
```diff
@@ -286,7 +286,8 @@ var (
 	}
-	reservedIdentityNames = map[NumericIdentity]string{
+	ReservedIdentityNames = map[NumericIdentity]string{
+		IdentityUnknown:            "unknown",
```
For reviewers: Given how we're using `ReservedIdentityNames`, it looks okay to add the unknown identity to the list (+ it is, in a sense, a reserved identity as well). If that's not okay, I can work around it in the new `Entity()` function.

joestringer replied:
At a glance it seems like it should be OK? 

I'm a little bit wary that this moves us more in the direction of defining what an "unknown" identity is, when it really represents a lack of information rather than something specific. If we had perfect information in all monitor output messages it wouldn't exist. But I don't think this change will particularly influence this.

:+1:

kkourt commented on pkg/monitor/api/types.go:
```diff
@@ -551,3 +552,11 @@ func (m PolicyMatchType) String() string {
+
+// Entity prints the ID in a human readable string
+func Entity(id uint32) string {
+	if err, ok := identity.ReservedIdentityNames[identity.NumericIdentity(id)]; ok {
```
nit: the `err` variable name seems somewhat confusing here.

kkourt reviewed (approved):
Looks good to me.

christarazi reviewed (approved):
LGTM.

joestringer commented on pkg/monitor/datapath_drop.go:
```diff
@@ -43,8 +43,8 @@ type DropNotify struct {
-	fmt.Printf("xx drop (%s) flow %#x to endpoint %d, identity %d->%d: %s\n",
-		api.DropReason(n.SubType), n.Hash, n.DstID, n.SrcLabel, n.DstLabel,
+	fmt.Printf("xx drop (%s) flow %#x to endpoint %d, identity %s->%s: %s\n",
+		api.DropReason(n.SubType), n.Hash, n.DstID, api.Entity(n.SrcLabel), api.Entity(n.DstLabel),
```
I wonder if it would still be useful to also have something like a `-n`, `--numeric` option to still support printing the numeric identities in case we find it easier to read the numbers?

pchaigno replied:
I implemented this as the last commit.
@brb Does it also address your concern?

joestringer commented on pkg/monitor/api/types.go:
```diff
@@ -551,3 +552,11 @@ func (m PolicyMatchType) String() string {
 }
+
+// Entity prints the ID in a human readable string
+func Entity(id uint32) string {
```
I think `Entity` as a naming here crosses two separate concepts and I don't think it's a good idea to mix them up. The input is a numeric Identity, which represents exactly one possible security context of the peer in the connection. An Entity on the other hand is a policy-level selector which covers multiple individual Identities (for example, `cluster` entity covers various different identities that are all conceptually part of the 'cluster').

I wonder whether we could just convert the uint32 from monitor side into a native security identity type instead and get the string conversion from there? I'm looking at this which seems to be equivalent to this function:

https://github.com/cilium/cilium/blob/91748c7e19b4afd3a7d30087b7ba5f69da25452f/pkg/identity/numericidentity.go#L364-L370

joestringer reviewed (changes requested):
Overall I think this is a net positive change. :+1: 

I think this is probably one of those things that will make it easier to understand the CLI for newer users, while oldtimers like me will grumble about not having the simple numeric identities to grep for & easily recognize ;)

The `Entity()` function is confusing and I think duplicated with an existing function, so we should address that. The other points below are food for thought.

pchaigno committed:
monitor: New --numeric flag to display numeric identities

The output is bufferized to avoid the monitor's info log ("level=info
msg="Initializing dissection cache..." subsys=monitor") from being
displayed in the middle of the first output line.

Suggested-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Paul Chaignon <paul@cilium.io>

kkourt commented on pkg/monitor/datapath_drop.go:
```diff
@@ -43,30 +45,43 @@ type DropNotify struct {
+	fmt.Fprintf(buf, "xx drop (%s) flow %#x to endpoint %d",
+		api.DropReason(n.SubType), n.Hash, n.DstID)
+	if numeric {
+		fmt.Fprintf(buf, ", identity %s->%d", n.SrcLabel, n.DstLabel)
```
Was this supposed to be `%d->%d` instead of `%s->%d`?

(A helper function for the 4 lines to print the identity might help here since we seem to be using it a couple of times)

kkourt reviewed (approved):
Modulo a minor comment bellow, LGTM.

Thanks on behalf of all of us who can't keep a lot of numbers on their head :-)

ti-mo reviewed (approved):
Thanks! This lowers the barrier to entry for newbies and is definitely a clarity improvement. Please stop assuming that people dropping in to hack on the code magically know what `0->1` means in the output of a CLI tool.

For science, the count of `#define .. 0`s:
```
λ  cilium master ✓ grep -R "#define.*\s0$" | wc -l
43
```

And that doesn't even include the occurrences where we initialize a variable to 0 and pass it to `send_trace_notify()` to result in an `unknown` popping up in a CLI (as is still the case here). I had to explicitly be told this. Yes, I'm saying we should add a `#define ID_UNKNOWN 0` to make this explicit. :wink:

twpayne commented on pkg/monitor/datapath_drop.go:
```diff
@@ -34,38 +37,53 @@ type DropNotify struct {
 
+// dumpIdentity dumps the source and destination identities in numeric or
+// human-readable format.
+func (n *DropNotify) dumpIdentity(buf *bufio.Writer, numeric bool) {
```
Tiny nit: `bool` parameters are hard to reason about when reading the callsite code. Calls will look like `n.dumpIdentity(buf, false)` and it's not clear what `false` means in this context.

Consider defining a type and using that:

```go
type DisplayFormat bool

const (
    DisplayLabel   DisplayFormat = false
    DisplayNumeric DisplayFormat = true
)

func (n *DropNotify) dumpIdentity(buf *bufio.Writer, df DisplayFormat) {
    // ...
}
```

This will make the callsite look like (for example) `n.dumpIdentity(buf, DisplayLabel)`.

twpayne commented on pkg/monitor/datapath_drop.go:
```diff
@@ -34,38 +37,53 @@ type DropNotify struct {
+// human-readable format.
+func (n *DropNotify) dumpIdentity(buf *bufio.Writer, numeric bool) {
+	if numeric {
+		fmt.Fprintf(buf, "identity %d->%d", n.SrcLabel, n.DstLabel)
```
(This comment applies elsewhere too)

`buf` can be the more generic `io.Writer` here, and elsewhere. Using the `io.Writer` interface as a destination rather than a specific type is used consistently throughout the codebase.

twpayne commented on pkg/monitor/datapath_drop.go:
```diff
@@ -34,38 +37,53 @@ type DropNotify struct {
+		api.DropReason(n.SubType), n.Hash, n.DstID)
+	n.dumpIdentity(buf, numeric)
+	fmt.Fprintf(buf, ": %s\n", GetConnectionSummary(data[DropNotifyLen:]))
+	buf.Flush()
```
This code hardwires `os.Stdout` as a destination into the function. The code will be more flexible and testable if instead an `io.Writer` is passed in and the caller choses the destination.

pchaigno replied:
I agree on the general idea, but I'm not sure it's worth it in this case. 1) It's not a regression introduced by the new code changes, and more importantly, 2) it's unlikely we'll ever unit test those functions because they are very simple. Passing the destination as an argument of the function will make the caller's code a bit more complex and will add yet another argument to those functions.

twpayne commented on pkg/monitor/datapath_drop.go:
```diff
@@ -34,38 +37,53 @@ type DropNotify struct {
-func (n *DropNotify) DumpVerbose(dissect bool, data []byte, prefix string) {
-	fmt.Printf("%s MARK %#x FROM %d DROP: %d bytes, reason %s",
+func (n *DropNotify) DumpVerbose(dissect bool, data []byte, prefix string, numeric bool) {
+	buf := bufio.NewWriter(os.Stdout)
```
Same comment as previous: pass in an `io.Writer` instead of hardwiring `os.Stdout`.

twpayne reviewed (commented):
This is a great functional improvement for newbies like me :) See individual comments on the use of Go, particularly avoiding `bool` parameters and using `io.Writer` as a destination.
